dali/kernels/reduce/reduce_axes_gpu_impl.cuh

```diff
@@ -17,38 +17,194 @@
 
 
 #include "dali/core/util.h"
+#include "dali/core/geom/vec.h"
 #include "dali/kernels/reduce/reductions.h"
 #include "dali/kernels/reduce/reduce_all_gpu_impl.cuh"
 #include "dali/kernels/reduce/reduce_common.cuh"
 #include "dali/kernels/reduce/online_reducer.h"
+#include "dali/kernels/reduce/reduce_drop_dims.h"
 
 namespace dali {
 namespace kernels {
 
+namespace reduce_impl {
+/**
+ *  A preprocessor bank should satisfy the concept:
+ *
+ *  ```
+ *  struct PreprocessorBank {
+ *      SingleValuePreprocessor Get(const i64vec<non_reduced_dims> &pos) const;
+ *  };
+ *  ```
+ *  where SingleValuePreprocessor is a unary functor.
+ */
+
+/**
+ * @brief A no-op preprocessor bank,
+ */
+template <int non_reduced_ndim>
+struct IdentityPreprocessor {
+  DALI_HOST_DEV DALI_FORCEINLINE
+  dali::identity Get(const i64vec<non_reduced_ndim> &) const {
+    return {};
+  }
+};
+
+/**
+ * @brief Calculates variance, given a tensor of means.
+ */
+template <int non_reduced_ndim, typename Mean>
+struct VariancePreprocessor;
+
+template <typename Mean>
+struct VariancePreprocessor<1, Mean> {
+  const Mean *mean;
+  i64vec<1> stride;
+
+  template <typename T>
+  DALI_HOST_DEV DALI_FORCEINLINE
+  reductions::variance<Mean> Get(const i64vec<1> &pos) const {
+    auto offset = dot(pos, stride);
+  #ifdef __CUDA_ARCH__
+    float m = __ldg(mean + offset);
```

Comment:
Do you need to use `__ldg` explicitly or compiler can do that for you?

Reply:
Either that, or `const Mean *__restrict__ mean`

Reply:
Sure